lint/rule.go

```diff
@@ -14,10 +14,304 @@
+	return r.Name
+}
+
+// GetURI returns the URI where the applicable guideline is documented.
+func (r *EnumRule) GetURI() string {
+	return r.URI
+}
+
+// Lint accepts a FileDescriptor and lints every enum in the file
+// (including enums nested within messages).
+func (r *EnumRule) Lint(fd *desc.FileDescriptor) []Problem {
+	problems := []Problem{}
+
+	// Lint enums that are at the top level of the file.
+	for _, enum := range fd.GetEnumTypes() {
+		problems = append(problems, r.LintEnum(enum)...)
+	}
+
+	// Lint enums that are nested within messages.
+	for _, message := range getAllMessages(fd) {
+		for _, enum := range message.GetNestedEnumTypes() {
+			problems = append(problems, r.LintEnum(enum)...)
+		}
+	}
+	return problems
+}
+
+// ruleIsEnabled returns true if the rule is enabled (not disabled by the comments
+// for the given descriptor or its file), false otherwise.
+func ruleIsEnabled(rule ProtoRule, d desc.Descriptor) bool {
+	directive := fmt.Sprintf("api-linter: %s=disabled", rule.GetName())
+
+	// If the comments above the descriptor disable the rule,
+	// return false.
+	if sourceInfo := d.GetSourceInfo(); sourceInfo != nil {
+		if strings.Contains(sourceInfo.GetLeadingComments(), directive) {
+			return false
+		}
+	}
+
+	// The rule may also be disabled at the file level.
+	// If it is, return false.
+	if strings.Contains(fileHeader(d.GetFile()), directive) {
+		return false
+	}
+
+	// The rule is enabled.
+	return true
+}
+
+// getAllMessages returns a slice with every message (not just top-level
+// messages) in the file.
+func getAllMessages(f *desc.FileDescriptor) (messages []*desc.MessageDescriptor) {
+	messages = append(messages, f.GetMessageTypes()...)
+	for _, message := range f.GetMessageTypes() {
+		messages = append(messages, getAllNestedMessages(message)...)
+	}
+	return messages
+}
+
+// getAllNestedMessages returns a slice with the given message descriptor as well
+// as all nested message descriptors, traversing to arbitrary depth.
+func getAllNestedMessages(m *desc.MessageDescriptor) (messages []*desc.MessageDescriptor) {
+	for _, nested := range m.GetNestedMessageTypes() {
+		messages = append(messages, nested)
+		messages = append(messages, getAllNestedMessages(nested)...)
+	}
+	return messages
+}
+
+// fileHeader attempts to get the comment at the top of the file, but it
+// is on a best effort basis because protobuf is inconsistent.
+//
+// Taken from https://github.com/jhump/protoreflect/issues/215
+func fileHeader(fd *desc.FileDescriptor) string {
+	var firstLoc *descriptor.SourceCodeInfo_Location
+	var firstSpan int64
+	for _, curr := range fd.AsFileDescriptorProto().GetSourceCodeInfo().GetLocation() {
+		// Skip locations that have no comments.
+		if curr.LeadingComments == nil && len(curr.LeadingDetachedComments) == 0 {
+			continue
+		}
+		// Skip locations that are not syntax, package, option, or import.
+		allowedPaths := map[int32]struct{}{2: {}, 3: {}, 8: {}, 12: {}}
```

Comment:
Bring it out of the for loop, we don't need to create it in every loop.

Reply:
Done.